dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalDAO.java

```
@@ -890,24 +893,103 @@ public <ASPECT extends RecordTemplate, RELATIONSHIP extends RecordTemplate> List
+
+    // if DUAL_SOURCE mode, verify that the relationships extracted from aspect and built from relationship builders have the same source/destination pairs,
+    // with different relationship types. Else, throw an exception.
+    if (areConsistentLocalRelationshipUpdates(updatesExtractedFromAspect, updatesBuiltFromRelationshipBuilders)) {
+      _localRelationshipWriterDAO.processLocalRelationshipUpdates(urn, updatesExtractedFromAspect, isTestMode);
+      _localRelationshipWriterDAO.processLocalRelationshipUpdates(urn, updatesBuiltFromRelationshipBuilders, isTestMode);
     } else {
-      throw new UnsupportedOperationException("Please ensure that the RelationshipSource enum is properly set using "
-          + "setRelationshipSource method.");
+      throw new IllegalArgumentException(String.format(
+          "Inconsistent relationships extracted from aspect and built from relationship builders from aspect: %s.", aspect));
+    }
+
+    // returns both updates from aspect and relationship builders
+    return Stream.concat(updatesExtractedFromAspect.stream(), updatesBuiltFromRelationshipBuilders.stream())
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Checks data conflicts between relationships extracted from aspect and relationships built from relationship builders.
+   * Conflict:
+   * 1. Both relationships are the same type.
+   * 2. Two relationships have different source/destination pairs.
+   * @param updatesExtractedFromAspect LocalRelationshipUpdates extracted from aspect
+   * @param updatesBuiltFromRelationshipBuilders LocalRelationshipUpdates built from relationship builders
+   * @return true if there is no conflict, false otherwise
+   */
+  private boolean areConsistentLocalRelationshipUpdates(List<LocalRelationshipUpdates> updatesExtractedFromAspect,
+      List<LocalRelationshipUpdates> updatesBuiltFromRelationshipBuilders) {
+    if (updatesExtractedFromAspect.isEmpty() || updatesBuiltFromRelationshipBuilders.isEmpty()) {
```

It is possible that this is true while there are still relationships extracted from the relationship builders. 
```
AspectFoo {
    bars = [b]
    belongsTos = []
}
```

The relationship builder can return {A->B} but belongsTos is empty. This should throw an error, not return true.